server/src/main/java/org/opensearch/common/time/DateTime.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.time;
+
+import java.time.DateTimeException;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoField;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Container class for parsed date/date-time data.
+ */
+class DateTime {
```

Comment:
Why do we need this class vs parsing into `ZonedDateTime` or `OffsetDateTime`? (and consequently, LocalDate + LocalTime + ZoneId/ZoneOffset)?

Reply:
We need this class mainly to make it easier for parsers to construct datetime. It provides specialised constructors which isn't available in classes like `ZonedDateTime` and `OffsetDateTime`. It abstracts out creation and transformation of datetime into relevant type like `LocalDateTime`, `OffsetDateTime` etc.

Reply:
I am not sure I get that - this class is LocalTime + LocalDate + ZoneOffset (the `fractionDigits` it not used anywhere) - all this classes are data bags that could be used without any restrictions. If we think about having more custom parsers, we should use classes from standard library whenever possible.

Reply:
@reta I think the pros of this data bag class outweighs the cons
Pros:
- Provides specialised constructors which makes the implementation cleaner for eg. [here](https://github.com/opensearch-project/OpenSearch/pull/11465/files#diff-a93d879c0229ef71b12bfd44347d9bad505fc8bcb6bb4d18ba98ab232ca99dd8R132) if we would've used the standard library constructors say `of(int year, int month, int dayOfMonth, int hour, int minute, int second, int nanoOfSecond, zone)` function of zonedDateTime then it'll require mentioning default values of all other params like seconds, nanos, zone etc. This will result in multiple instances where we need to provide these default values across the parser implementation making the code bloated
- Provides application level [sanity checks](https://github.com/opensearch-project/OpenSearch/pull/11465/files#diff-e25ec903a0c1536ed62135b3bf331efa2977a1b9753722901980b99e482931d0R42) on the parts of datetime
- It contains all the parts of datetime (others can be easily added if required by providing specialised constructor) and can be easily converted into the required datetime type like `LocalDateTime`, `offSetDateTime` etc. for eg. local datetime parser just needed one conversion function to get the required datetime type [here](https://github.com/opensearch-project/OpenSearch/pull/11675/files#diff-e25ec903a0c1536ed62135b3bf331efa2977a1b9753722901980b99e482931d0R168)

Cons:
- Data bag class which can be used without any restrictions
- Poses slight stateful overhead

Let me know if you think we can keep it given its pros
PS: I've removed `fractionDigits`

Reply:
@CaptainDredge every piece of code we are adding adds the maintenance cost. This class (in my opinion) is not helpful and in fact, is unsafe since it could blow at runtime by throwing exception sporadically (please check `OffsetDateTime toOffsetDatetime` for example). 

If you really think this class is necessary, make it a private member of `RFC3339DateTimeFormatter` so it won't escape anywhere.

Reply:
Sure @reta I get the maintenance overhead point and hence removed it entirely